### PR TITLE
feat(zk): add shuffle argument

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements [shuffle](https://github.com/privacy-scaling-explorations/halo2/tree/main/halo2_backend/src/plonk/shuffle) argument.

### CODE REVIEW

File: tachyon/zk/plonk/vanishing/BUILD.bazel [hunk 1]
```diff
@@ -26,13 +26,14 @@ tachyon_cc_library(
         "//tachyon/base/types:always_false",
         "//tachyon/zk/base:rotation",
         "//tachyon/zk/lookup/halo2:evaluator",
-        "//tachyon/zk/lookup/halo2:prover",
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/vanishing/BUILD.bazel:29]: Is this change relevant to this commit?

File: tachyon/zk/plonk/expressions/evaluator/cells_querier.h [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+          query.SetIndex(cells_.meta_->QueryInstanceIndex(query.column(),
+                                                          query.rotation()));
+        }
+        return;
+      }
+      case ExpressionType::kChallenge:
+        return;
+      case ExpressionType::kSelector: {
+        Selector selector = input->ToSelector()->selector();
+        // TODO(chokobole): should it be std::set<Selector>?
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/expressions/evaluator/cells_querier.h:75]: For `queried_selectors_`?

**Comment by TomTaehoonKim** [comment]
Could you add a link for the test in [93a9086](https://github.com/kroma-network/tachyon/pull/488/commits/93a90861266112cc0dd26d37c59c0ab97ef98595) please?

File: tachyon/zk/plonk/halo2/config.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+#include <stdint.h>
+
+#include "tachyon/export.h"
+#include "tachyon/zk/plonk/halo2/vendor.h"
+
+namespace tachyon::zk::plonk::halo2 {
+
+struct TACHYON_EXPORT Config {
+  // By default, halo2 behaves like scroll halo2 v1.1.
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/zk/plonk/halo2/config.h:12]: This one https://github.com/scroll-tech/halo2/tree/v1.1

File: tachyon/zk/plonk/vanishing/BUILD.bazel (hunk 1 above)
Comment (chokobole) [reply tachyon/zk/plonk/vanishing/BUILD.bazel:29]: Nope should be in separate commit, I'll fix this!

File: tachyon/zk/plonk/expressions/evaluator/cells_querier.h (hunk 2 above)
Comment (chokobole) [reply tachyon/zk/plonk/expressions/evaluator/cells_querier.h:75]: Yes!

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/zk/shuffle/prover.h [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  template <typename Domain>
+  static Pair<Evals> CompressPair(
+      const Domain* domain, const Argument<F>& argument, const F& theta,
+      const plonk::ProvingEvaluator<Evals>& evaluator_tpl);
+
+  template <typename PCS>
+  static BlindedPolynomial<Poly, Evals> CreateGrandProductPoly(
+      ProverBase<PCS>* prover, const Pair<Evals>& compressed_pair,
+
+      const F& gamma);
```
Comment (GideokKim) [inline tachyon/zk/shuffle/prover.h:106]:
```suggestion
      ProverBase<PCS>* prover, const Pair<Evals>& compressed_pair,
      const F& gamma);
```

File: tachyon/zk/shuffle/evaluator.h [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+      shuffle_evaluators_.push_back(std::move(graph_input));
+      shuffle_evaluators_.push_back(std::move(graph_shuffle));
+    }
+  }
+
+  template <typename PCS, typename LS>
+  void UpdateChunkByShuffles(plonk::CircuitPolynomialBuilder<PCS, LS>& builder,
+                             absl::Span<F> chunk, size_t chunk_offset,
+                             size_t chunk_size) {
+    for (size_t i = 0; i < shuffle_evaluators_.size(); i += 2) {
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/zk/shuffle/evaluator.h:76]: Nice catch, but then range check should be changed to `shuffle_evaluators_.size() / 2`

**Review by GideokKim — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by batzor — APPROVED** [review]
LGTM
